### PR TITLE
trim PID

### PR DESCRIPTION
**Pull request by pzinn**
- correct trimming of modules / ideals in PIDs. resolves (at last!) #1273:
```
i1 : R = ZZ/7[t];

i2 : I = ideal(t^3*(t+1), t*(t-1))

             4    3   2
o2 = ideal (t  + t , t  - t)

o2 : Ideal of R

i3 : trim I

o3 = ideal t

o3 : Ideal of R

i4 : R=QQ[x]

o4 = R

o4 : PolynomialRing

i5 : ideal(3*x^4-x^3+x^2-x-2,3*x^4+2*x^3+6*x+4)

              4    3    2            4     3
o5 = ideal (3x  - x  + x  - x - 2, 3x  + 2x  + 6x + 4)

o5 : Ideal of R

i6 : trim oo

               2
o6 = ideal(x + -)
               3

o6 : Ideal of R
```
- several bugs were found in the process. #3017 is *not* fixed in this PR, somebody else needs to take a look at it since a lot is problematic with `smithNormalForm` at the moment.
- instead, a temporary workaround in the `prune` PID strategy was implemented, and similarly for the new `trim` PID strategy.
- tests were added for both `trim` and `prune`.

### CODE REVIEW

File: M2/Macaulay2/tests/normal/trim.m2 [hunk 1]
```diff
@@ -25,6 +25,13 @@ assert (trim(M,Strategy=>Complement) === trim(N,Strategy=>Complement))
 assert (trim P === R^1)
 assert (trim(P,Strategy=>Complement) === R^1)
 
+-- trim PID
+R=QQ[x]
+I = ideal(3*x^4-x^3+x^2-x-2,3*x^4+2*x^3+6*x+4)
+J = trim I
+assert(I == J)
+assert(numgens J == 1)
```
Comment (mahrud) [inline M2/Macaulay2/tests/normal/trim.m2:33]: Here's an example that currently fails, if you'd like to use different examples:
```m2
i1 : R = QQ[x]

o1 = R

o1 : PolynomialRing

i2 : I = ideal(x^3-1, x^2+1)

             3       2
o2 = ideal (x  - 1, x  + 1)

o2 : Ideal of R

i3 : trim I
stdio:3:1:(3): error: matrix column index 1 out of range 0 .. 0
```

File: M2/Macaulay2/m2/matrix2.m2 [hunk 2]
```diff
@@ -247,6 +247,26 @@ trimHelper = ((opts, M) -> (
...
+    p:=id_(target ch)//ch;
+    q:=generators M*p;
+    d:=diagonalMatrix apply(rank source q,i->lcm apply(entries (q_i),a->if a==0 then 1 else 1/leadCoefficient a));
+    image (q*d)
+    )
+addHook((trim, Module), Strategy => "PID",
+    (opts, M) -> (
+	R := ring M;
+	if instance(R,PolynomialRing) and numgens R === 1 and isField coefficientRing R and not isHomogeneous M then trimPID M
+	)
```
Comment (mahrud) [inline M2/Macaulay2/m2/matrix2.m2:266]: Rather than repeating the prune algorithm in trimPID, how about this instead?
```suggestion
	then if isFreeModule M then M else subquotient(
	    if M.?generators   then (prune image generators M).cache.pruningMap,
	    if M.?relations    then (prune image relations  M).cache.pruningMap))
```

Comment (pzinn) [reply M2/Macaulay2/m2/matrix2.m2:266]: I don't think what you wrote works. anyway I did have an earlier version going through `prune`, but it was fairly complicated. I think the current code is cleaner.

File: M2/Macaulay2/tests/normal/trim.m2 (hunk 1 above)
Comment (pzinn) [reply M2/Macaulay2/tests/normal/trim.m2:33]: weird, works just fine here.

File: M2/Macaulay2/m2/matrix2.m2 (hunk 2 above)
Comment (mahrud) [reply M2/Macaulay2/m2/matrix2.m2:266]: Sorry, what doesn't work?

Comment (pzinn) [reply M2/Macaulay2/m2/matrix2.m2:266]: I tried your code with `coker matrix {{x}}` and it didn't produce the expected result.

Comment (mahrud) [reply M2/Macaulay2/m2/matrix2.m2:266]: What's the expected result, not this?!
```m2
i1 : R = QQ[x]; trim coker matrix {{x}}

o2 = cokernel | x |
```

Comment (mahrud) [reply M2/Macaulay2/m2/matrix2.m2:266]: You must have used `M.relations` for both inputs of `subquotient`.

Comment (pzinn) [reply M2/Macaulay2/m2/matrix2.m2:266]: no, I copy/pasted your code:
```
i1 : test = M -> if isFreeModule M then M else subquotient(
	    if M.?generators   then (prune image generators M).cache.pruningMap,
	    if M.?relations    then (prune image relations  M).cache.pruningMap)
          
o1 = test

o1 : FunctionClosure

i2 : R=QQ[x];

i3 : test cokernel matrix {{x}}

o3 = subquotient (| x |, | x |)

                               1
o3 : R-module, subquotient of R
```

Comment (pzinn) [reply M2/Macaulay2/m2/matrix2.m2:266]: also it's about twice as slow due lots of junk in the `prune` code.

Comment (mahrud) [reply M2/Macaulay2/m2/matrix2.m2:266]: These are valid improvements to the PID strategy for prune that I think should be implemented, but the main computation at the center of trim and prune for an ideal or module over PID is exactly the same, and duplicating an algorithm in two places is a recipe for bugs and inconsistencies and harder maintenance in the future.

If you want me to suggest a different change that doesn't have the slow down you describe (pull the prune strategy into a helper function and use it both places) or the denominators (divide by the leading coefficients) I can, but this is your PR and since my impression is that you also don't like hard to parse code I wanted to suggest a way to avoid this mess:
```m2
trimPID := M -> if M.?relations then (if M.?generators then trimPID image generators M else ambient M) / trimPID image relations M else if not M.?generators then M else (
```

If you don't care about code duplication and inconsistency between trim and prune then never mind.

Comment (pzinn) [reply M2/Macaulay2/m2/matrix2.m2:266]: I'll try to rewrite to avoid redundancies and ifs
